crates/qasm3/src/lib.rs

```diff
@@ -0,0 +1,207 @@
+mod build;
+mod circuit;
+mod error;
+mod expr;
+
+use std::ffi::OsString;
+use std::path::{Path, PathBuf};
+
+use hashbrown::HashMap;
+
+use pyo3::prelude::*;
+use pyo3::types::{PyModule, PyTuple};
+
+use oq3_semantics::syntax_to_semantics::parse_source_string;
+
+use crate::error::QASM3ImporterError;
+
+/// The name of a Python attribute to define on the given module where the default implementation
+/// of the ``stdgates.inc`` custom instructions is located.
+const STDGATES_INC_CUSTOM_GATES_ATTR: &str = "STDGATES_INC_GATES";
+
+/// Load an OpenQASM 3 program from a string into a :class:`.QuantumCircuit`.
+///
+/// Args:
+///     source (str): the program source in a Python string.
+///     custom_gates (Iterable[CustomGate]): Python constructors to use for particular named gates.
+///         If not supplied, Qiskit will use its own standard-library constructors for gates
+///         defined in the OpenQASM 3.0 standard-library file ``stdgates.inc``.
+///     include_path (Iterable[str]): the path to search when resolving ``include`` statements.
+///         If not given, Qiskit will arrange for this to point to a location containing
+///         ``stdgates.inc`` only.  Paths are tried in the sequence order.
+///
+/// Returns:
+///     :class:`.QuantumCircuit`: the constructed circuit object.
+///
+/// Raises:
+///     :class:`.QASM3ImporterError`: if an error occurred during parsing or semantic analysis.
+///         In the case of a parsing error, most of the error messages are printed to the terminal
+///         and formatted, for better legibility.
+#[pyfunction]
+#[pyo3(pass_module, signature = (source, /, *, custom_gates=None, include_path=None))]
+pub fn loads(
+    module: &PyModule,
+    py: Python,
+    source: String,
+    custom_gates: Option<Vec<circuit::PyGate>>,
+    include_path: Option<Vec<OsString>>,
+) -> PyResult<circuit::PyCircuit> {
+    let default_include_path = || -> PyResult<Vec<OsString>> {
+        Ok(vec![Path::new(module.filename()?)
+            .parent()
+            .unwrap()
+            .join(["qasm", "libs", "dummy"].iter().collect::<PathBuf>())
+            .into_os_string()])
+    };
+    let include_path = include_path.map(Ok).unwrap_or_else(default_include_path)?;
+    let result = parse_source_string(source, None, Some(&include_path));
+    if result.any_errors() {
+        result.print_errors();
+        return Err(QASM3ImporterError::new_err(
+            "errors during parsing; see printed errors",
+        ));
+    }
+    let gates = match custom_gates {
+        Some(gates) => gates
+            .into_iter()
+            .map(|gate| (gate.name().to_owned(), gate))
+            .collect(),
+        None => module
+            .getattr(STDGATES_INC_CUSTOM_GATES_ATTR)?
+            .iter()?
+            .map(|obj| {
+                let gate = obj?.extract::<circuit::PyGate>()?;
+                Ok((gate.name().to_owned(), gate))
+            })
+            .collect::<PyResult<HashMap<_, _>>>()?,
```

Comment:
Is there a model where we want to be able to extend the standard gate definitions with custom ones? Right now it's either or.

Reply:
Right now it's possible to do - I exposed `qiskit.qasm3.STDGATES_INC_GATES`, so you can copy that and append to it.  Are you thinking of a different interface?

Reply:
I stump again here for extern gate declarations. 

If you have `extern gate h q;`, then the parser doesn't error when it finds a call to `h`. And the importer knows that it can use its standard definition of `h` . (or whatever qiskit-specific override mechanism exists).

But if you write `gate h q {... contents ..}`, then this unambiguously overrides Qiskit builtins.

Reply:
John: it's an interestoing language point, but it doesn't affect the issue at hand here, and it's a matter for the OQ3 design committee, not us in our roles on Qiskit.

This attribute _is_ the "qiskit-specific override mechanism" you describe - it doesn't matter how a "gate" symbol appears in the symbol table, this system is for specifying how to relate gate entries to Python-space constructors, with type checking for the signature (that's why the `num_params` and `num_qubits` are given).